content/3_Silver/Graph_Traversal.mdx

```diff
@@ -247,6 +166,286 @@ d.pop()  # [1, 3]
+
+```cpp
+const int MAX_N = 1e6;
+vector<int> adj[MAX_N];
+bool visited[MAX_N];
+
+void dfs(int current_node) {
+	if (visited[current_node]) { return; }
+	visited[current_node] = true;
+
+	for (int neighbor : adj[current_node]) { dfs(neighbor); }
+}
+
+int main() {
+	// Define adjacency list and read in problem-specific input
+
+	for (int i = 0; i < MAX_N; i++) {
+		// iterate over all connected components in the graph
+		if (!visited[i]) { dfs(i); }
+	}
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+
+#### BFS
+
+When implementing BFS, we often use a queue to track the next vertex to visit. Like DFS, we'll also keep an array to store whether we've seen a vertex before.
+
+The code below shows the variables we need to maintain and an iterative implementation of DFS.
+
+<LanguageSection>
+<JavaSection>
+
+```java
+static final int MAX_N = (int)1e6;
+
+public static void main(String[] args) throws IOException {
+	boolean[] visited = new boolean[MAX_N];
+	ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>();
+	for (int i = 0; i < MAX_N; i++) adj.add(new ArrayList<Integer>());
+
+	// Define adjacency list and read in problem-specific input
+
+	for (int i = 0; i < MAX_N; i++) {
+		// iterate over all connected components in the graph
+		if (!visited[i]) {
+			Queue<Integer> q = new LinkedList<Integer>();
+			q.add(i);
+
+			while (q.size() > 0) {
+				int currentNode = q.peek();
+				q.poll();
+
+				if (visited[currentNode]) { continue; }
+				visited[currentNode] = true;
+
+				for (int neighbor : adj.get(currentNode)) { q.add(neighbor); }
+			}
+		}
+	}
+}
+```
+
+</JavaSection>
+<CPPSection>
+
+```cpp
+const int MAX_N = 1e6;
+
+int main() {
+	vector<int> adj[MAX_N];
+	bool visited[MAX_N] = {0};
+
+	// Define adjacency list and read in problem-specific input
+
+	for (int i = 0; i < MAX_N; i++) {
+		// iterate over all connected components in the graph
+		if (!visited[i]) {
+			queue<int> q;
+			q.push(i);
+
+			while (!q.empty()) {
+				int current_node = q.front();
+				q.pop();
+
+				if (visited[current_node]) { continue; }
+				visited[current_node] = true;
+
+				for (int neighbor : adj[current_node]) { q.push(neighbor); }
+			}
+		}
+	}
+}
+```
+
+</CPPSection>
+<PySection>
+
+```py
+from collections import deque
+
+MAX_N = int(1e6)
+adj = [[] for _ in range(MAX_N)]
+visited = [False] * MAX_N
+
+# Define adjacency list and read in problem-specific input
+
+for i in range(MAX_N):
+	# iterate over all connected components in the graph
+	if not visited[i]:
+		q = deque([i])
+
+		while q:
+			node = q.popleft()
+			visited[node] = True
+
+			for neighbor in adj[node]:
+				if not visited[neighbor]:
+					q.append(neighbor)
+```
+
+</PySection>
+</LanguageSection>
+
+<Resources>
+	<Resource source="CPH" title="12 - Graph traversal" starred />
+</Resources>
+
+
+## Applications
+
+### Connected Components
+
+<FocusProblem problem="connSample" />
+
+A **connected component** is a maximal set of connected nodes in an undirected
```

Comment:
maximal?

Reply:
i didn't write this, but i believe that's the technical definition :/

let me know if you think this should be changed.

Reply:
ok idc then lol